stdlib/public/core/Span/Span.swift

```diff
@@ -0,0 +1,827 @@
+  @_alwaysEmitIntoClient
+  public func _elementsEqual(_ other: some Sequence<Element>) -> Bool {
+    var offset = 0
+    for otherElement in other {
+      if offset >= count { return false }
+      if self[unchecked: offset] != otherElement { return false }
+      offset += 1
+    }
+    return offset == count
+  }
+}
+
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 9999, *)
+extension Span where Element: ~Copyable {
+
+  /// The number of elements in the span.
+  ///
+  /// To check whether the span is empty, use its `isEmpty` property
+  /// instead of comparing `count` to zero.
+  ///
+  /// - Complexity: O(1)
+  @_alwaysEmitIntoClient
+  public var count: Int { _count }
+
+  /// A Boolean value indicating whether the span is empty.
+  ///
+  /// - Complexity: O(1)
+  @_alwaysEmitIntoClient
+  public var isEmpty: Bool { _count == 0 }
+
+  /// The representation for a position in `Span`.
+  public typealias Index = Int
+
+  /// The indices that are valid for subscripting the span, in ascending
+  /// order.
+  ///
+  /// - Complexity: O(1)
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  public var indices: Range<Index> {
+    Range(_uncheckedBounds: (0, _count))
+  }
+}
+
+//MARK: integer offset subscripts
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 9999, *)
+extension Span where Element: ~Copyable {
+
+  /// Accesses the element at the specified position in the `Span`.
+  ///
+  /// - Parameter position: The offset of the element to access. `position`
+  ///     must be greater or equal to zero, and less than `count`.
+  ///
+  /// - Complexity: O(1)
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  public subscript(_ position: Index) -> Element {
+    _read {
+      _precondition(indices.contains(position), "index out of bounds")
+      yield self[unchecked: position]
+    }
+  }
+
+  /// Accesses the element at the specified position in the `Span`.
+  ///
+  /// This subscript does not validate `position`; this is an unsafe operation.
+  ///
+  /// - Parameter position: The offset of the element to access. `position`
+  ///     must be greater or equal to zero, and less than `count`.
+  ///
+  /// - Complexity: O(1)
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @unsafe
+  @_alwaysEmitIntoClient
+  public subscript(unchecked position: Index) -> Element {
+    //FIXME: change to unsafeRawAddress or unsafeAddress when ready
+    _read {
+      let element = _start.advanced(by: position&*MemoryLayout<Element>.stride)
+      let binding = Builtin.bindMemory(element._rawValue, count._builtinWordValue, Element.self)
+      defer { Builtin.rebindMemory(element._rawValue, binding) }
+      yield UnsafePointer<Element>(element._rawValue).pointee
+    }
+  }
+}
+
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 9999, *)
+extension Span where Element: BitwiseCopyable {
```

Comment:
Is BitwiseCopyable the right constraint for this? Wouldn’t plain Copyable be OK?

Reply:
Oh nice. We hoped to be able to have unaligned spans for BitwiseCopyable types, but since that doesn't work `Copyable` would be fine here.

Reply:
After checking, `BitwiseCopyable` is still the right thing: eventually we _should_ be able to have unaligned spans for `BitwiseCopyable` types, so we want to use `loadUnaligned()` in the implementation of this subscript.